lib/node_modules/@stdlib/math/base/special/hyp2f1/lib/main.js

```diff
@@ -0,0 +1,284 @@
+	}
+
+	if ( a <= 0 && abs( a - ia ) < EPS ) { // a is negative integer
+		negIntA = 1;
+	}
+
+	if ( b <= 0 && abs( b - ib ) < EPS ) { // b is negative integer
+		negIntB = 1;
+	}
+
+	if ( d <= -1 && !( abs( d - id ) > EPS ) ) {
+		if ( s < 0 && !( negIntA || negIntB ) ) {
+			return pow( s, d ) * hyp2f1( c-a, c-b, c, x );
+		}
+	}
+
+	if ( d <= 0 && x === 1 && !( negIntA || negIntB )) {
+		// CEPHES: goto hypdiv
+		return PINF;
+	}
+	if ( ax < 1.0 || x === -1.0 ) { // 2F1(a,b;b;x) = (1-x)**(-a)
+		if ( abs( b-c ) < EPS ) { // b = c
+			if ( negIntB ) {
+				y = hyp2f1NegCEqualBC( a, b, x );
+			} else {
+				y = pow( s, -a );
+			}
+			// CEPHES: goto hypdon
+			return y;
+		}
+		if ( abs( a - c ) < EPS ) { // a = c
+			y = pow( s, -b );
+			return y;
+		}
+	}
+	if ( c <= 0.0 ) {
+		ic = round( c );
+		if ( abs( c - ic ) < EPS ) { // c is negative integer
+			// Check termination before explosion
+			if ( negIntA && ( ia > ic ) ) {
+				// CEPHES: goto hypok
+				y = hyt2f1( a, b, c, x, err );
+				return y;
+			}
+			if ( negIntB && ( ib > ic ) ) {
+				// CEPHES: goto hypok
+				y = hyt2f1( a, b, c, x, err );
+				return y;
+			}
+			// CEPHES: goto hypdiv
+			return PINF;
+		}
+	}
+
+	if ( negIntA || negIntB ) { // function in polynomial
+		// CEPHES: goto hypok
+		y = hyt2f1( a, b, c, x, err );
+		return y;
+	}
+	t1 = abs( b - a );
+	if ( x < -2.0 && abs( t1 - round(t1) ) > EPS ) {
+		/* This transform has a pole for b-a integer, and
+		 * may produce large cancellation errors for |1/x| close
+		 * to 1
+		 */
+		p = hyp2f1(	a, 1 - c + a, 1 - b + a, 1.0 / x );
+		q = hyp2f1(	b, 1 - c + b, 1 - a + b, 1.0 / x );
+		p *= pow( -x, -a );
+		q *= pow( -x, -b );
+		t1 = gamma( c );
+		s = t1 * gamma( b - a ) / (gamma( b ) * gamma( c - a ) );
+		y = t1 * gamma( a - b ) / (gamma( a ) * gamma( c - b ) );
+		return ( s * p ) + ( y * q );
+	}
+	if ( x < -1.0 ) {
+		if ( abs( a ) < abs( b ) ) {
+			return pow( s, -a ) * hyp2f1( a, c-b, c, x / ( x-1 ) );
+		}
+		return pow( s, -b ) * hyp2f1( b, c-a, c, x / ( x-1 ) );
+	}
+	if ( ax > 1.0 ) { // series diverges
+		// CEPHES: goto hypdiv
+		return PINF;
+	}
+
+	p = c - a;
+	ia = round( p ); // Nearest integer to c-a
+	if ( ( ia <= 0.0 ) && ( abs( p - ia ) < EPS ) ) { // negative int c-a
+		negIntCaOrCb = 1;
+	}
+
+	r = c - b;
+	ib = round( r ); // Nearest integer to c-b
+	if ( ( ib <= 0.0 ) && ( abs( r - ib ) < EPS ) ) { // negative int c-b
+		negIntCaOrCb = 1;
+	}
+
+	id = round( d );
+	q = abs( d - id );
+
+	if ( abs( ax - 1.0 ) < EPS ) { // |x| = 1.0
+		if ( x > 0.0 ) {
+			if ( negIntCaOrCb === 1 ) {
+				if ( d >= 0.0 ) { // goto hypf
+					y = pow( s, d ) * hys2f1( c-a, c-b, c, x, err );
+					return y;
+				}
+				// CEPHES: goto hypdiv
+				return PINF;
+			}
+			if ( d <= 0.0 ) { // goto hypdiv
+				return PINF;
+			}
+			y = gamma( c ) * gamma( d ) / ( gamma( p ) * gamma( r ) );
+			return y;
+		}
+		if ( d <= -1.0 ) {
+			return PINF;
+		}
+		if ( d < 0.0 ) {
+			y = hyt2f1( a, b, c, x, err );
```

Comment:
Another way of doing this API is to provide `hyt2f1` an output array as the first argument.

```javascript
hyt2f1( out, a, b, c, x );
y = out[ 0 ];
err = out[ 1 ];
```

As we do not need to worry about thread safety in JavaScript, the `out` array could be a "scratch" array initialized in the top-level scope, which is reused throughout this function whenever calling into one of these (synchronous) functions, such as `hyt2f1`. This would avoid the need for the function to create internal temporary objects, such as `err` (which adds to GC pressure), and also matches API conventions used elsewhere in the project.

Reply:
That makes a lot of sense.  I definitely want this to be consistent with previous work.  I will work on migrating that code over.

Reply:
Yeah, after starting on the rework, it seems like this should work.